### PR TITLE
Hw 29

### PR DESCRIPTION
**Pull request by tkhambir**

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 1]
```diff
@@ -32,6 +34,11 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
         String passwordRepeat = req.getParameter("pwd-repeat");
         if (password.equals(passwordRepeat)) {
             User user = new User(name, login, password);
+            if (user.getLogin().contains("admin")) {
+                user.setRoles(Set.of(Role.of("ADMIN")));
+            } else {
+                user.setRoles(Set.of(Role.of("USER")));
+            }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: first, admin cannot have a shopping cart
also, the user can choose a login that contains `admin`

let's create the admin user in InjectDataController instead
this controller may be accessed with a button from the main menu, for example

File: src/main/java/com/internet/shop/model/Role.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.internet.shop.model;
+
+public class Role {
+    private Long id;
+    private RoleName roleName;
+
+    public Role(RoleName roleName) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Role.java]:
```suggestion
    private Role(RoleName roleName) {
```

File: src/main/webapp/WEB-INF/views/users/accessDenied.jsp [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Title</title>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/users/accessDenied.jsp]: add informative title

File: src/main/webapp/web.xml [hunk 4]
```diff
@@ -184,4 +184,24 @@
...
+
+    <filter>
+        <filter-name>authorizationFilter</filter-name>
+        <filter-class>com.internet.shop.web.filters.AuthorizationFilter</filter-class>
+    </filter>
+
+    <filter-mapping>
+        <filter-name>authorizationFilter</filter-name>
+        <url-pattern>/*</url-pattern>
+    </filter-mapping>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]: let's maybe place controllers first, and then filters

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthorizationFilter implements Filter {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:20]: let's also refactor AuthenticationFilter as proposed in the checklist

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        if (isAuthorized(user, protectedUrls.get(requestedUrl))) {
+            chain.doFilter(req, resp);
+        } else {
+            req.getRequestDispatcher("/WEB-INF/views/users/accessDenied.jsp").forward(req, resp);
+        }
+    }
+
+    @Override
+    public void destroy() {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: redundant empty line

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String requestedUrl = req.getServletPath();
+
+        if (protectedUrls.get(requestedUrl) == null) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: we can also use `containsKey()` here

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    private final Map<String, List<Role.RoleName>> protectedUrls = new HashMap<>();
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        protectedUrls.put("/users", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/admin/products", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/orders/delete", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/admin/orders", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/users/delete", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products/add", List.of(Role.RoleName.ADMIN));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:33]: what about DeleteProductController?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Review by KseniiaMakarova — COMMENTED** [review]
it would be nice to have travis configured :)

**Pull request merged**